composition-js/src/validate.ts

```diff
@@ -306,9 +339,78 @@ export class ValidationState {
     }
-    return new ValidationState(newPath, newSubgraphPaths);
+
+    const updatedState = new ValidationState(newPath, newSubgraphPaths);
+
+    // We also ensure that all options have the same set or possible runtime types. If that's not the case, this
+    // would essentially mean that we can resolve a shared field from multiple subgraphs, but not all those subgraphs
+    // have the same possible runtime implementations for that field return type. But as a shardd field *must* be
+    // resolved the same way in all subgraphs, this is a red flag. Let's illustrate with an example:
+    // - subgraph S1:
+    //     Query {
+    //       e: E!
+    //     }
+    //
+    //     type E @key(fields: "id") {
+    //       id: ID!
+    //       s: U! @shareable
+    //     }
+    //
+    //     union U = A | B
+    //
+    //     type A {
+    //       a: Int
+    //     }
+    //
+    //     type B {
+    //       b: Int
+    //     }
+    // - subgraph S2:
+    //     type E @key(fields: "id") {
+    //       id: ID!
+    //       s: U! @shareable
+    //     }
+    //
+    //     union U = A | C
+    //
+    //     type A {
+    //       a: Int
+    //     }
+    //
+    //     type C {
+    //       c: Int
+    //     }
+    // In this case, both subgraph should only ever resolve `s` into an object of type `A`. Because if S1 resolves `s` into an object of
+    // type B, there is no way S2 could resolve it "the same way" (and vice-versa for S2 resolving `s` into a `C` object).
+    // But given the return type of `s` is `U` in both case, this feel like something that is easy to get wrong.
+    // And so we reject this example. To be clear, this is a somewhat conservative rejection: it *is* possible to implement this example
+    // correctly (by only ever resolving `s` into a `A` object) and users should be aware that shareable field should resolve the
+    // same way in all the subgraphs where they are resolved. But allowing it feels fairly error prone and it's not that clear that it
+    // is that useful in practice, so we don't allow it for now.
+    //
+    // Note that if motivated user demand for allowing the example above arises and we're not worried about users getting this wrong, then
+    // we could relax this rule. But even then, we should probably at least reject the case where the sets of runtime types for all
+    // the sources don't intersect, since there would be no good way to implement such case.
+    if (newSubgraphPaths.length > 1 && isAbstractType(newSubgraphPaths[0].tail.type)) {
+      const runtimeTypesToSubgraphs = new MultiMap<string, string>();
+      for (const path of newSubgraphPaths) {
+        // Note: we add spacing between elements because we're going use it if we display an error. This doesn't impact our set equality though
+        // since type names can have neither comma or space in graphQL.
+        runtimeTypesToSubgraphs.add(possibleRuntimeTypeNamesSorted(path).join(', '), path.tail.source);
```

Comment:
Trying to follow this, but I'm not sure if I'm getting it right. For the following example:
```
# subgraph 1
Query {
  a: A!
}

interface A {
  x: Int
}
type B implements A {
  x: Int
}

#subgraph 2
Query {
  a: A!
}

interface A {
  x: Int
}
type C implements A {
  x: Int
}
```
The things that are being compared are the internals of types B and C to make sure they match, so the string you're building will look something like "x:Int"? Is that right? If so, I think it looks good, if otherwise I may have more questions.

Reply:
So no. In that example, that code path will kick in after we've look at field `Query.a`, and we will essentially have 2 "options" (2 `newSubgraphPaths`), one per subgraph, but each one "tail" will be on interface `A`. So for the 1st path, `possibleRuntimeTypeNamesSorted` will return `B`, while  it will return `C`. And because those are different, we'll return the error. Does that help?